### PR TITLE
[SD-375] bypass tfa during pass reset use prlp

### PR DESCRIPTION
**Pull request by MdNadimHossain**
### Jira
https://digital-vic.atlassian.net/browse/SD-375
https://digital-vic.atlassian.net/browse/SD-421
### Problem/Motivation
The password reset workflow in our SDP projects, managed by the PRLP module, is being deprioritized due to a route subscriber defined in the TFA module. This conflict is disrupting the current password reset process for SDP projects.
### Fix
1. Override the TFA method and introduce a new route subscriber so that the password reset workflow is correctly handled by the PRLP module.
2. If the PRLP module is not enabled, fall back to the default password reset functionality provided by the TFA module.
3. By default, disable TFA for users with the previewer and secure_file_user roles.

### CODE REVIEW

**Comment by anthony-malkoun** [comment]
Why does this still have circleCI builds in it? Should the project be disconnected from Circle? Also, I don't think the failing GHA items ever worked?

**Review by vincent-gao — CHANGES_REQUESTED** [review]
hey @MdNadimHossain 
thanks for the changes
1. If possible, please use Dependency Injection to inject `prlpController` instead of directly instantiating it. This approach enhances modularity and testability.
2. `$request ? $request->getSession()->set('pass_reset_' . $uid, $token) : $_SESSION['pass_reset_' . $uid] = $token;` can be simplified to `$request->getSession()->set('pass_reset_' . $uid, $token);`. This change aligns with Drupal’s session handling practices, ensuring session consistency and better compatibility with symfony's request stack
3. If convenient, please replace `$request = \Drupal::request()` with `$this->requestStack->getCurrentRequest()`. Using the injected `RequestStack` service follows Drupal’s dependency injection practices and improves maintainability

**Comment by vincent-gao** [comment]
My comments don’t prevent this PR from being moved to testing.

**Comment by MdNadimHossain** [comment]
> * `$request ? $request->getSession()->set('pass_reset_' . $uid, $token) : $_SESSION['pass_reset_' . $uid] = $token;` can be simplified to `$request->getSession()->set('pass_reset_' . $uid, $token);`. This change aligns with Drupal’s session handling practices, ensuring session consistency and better compatibility with symfony's request stack
> * If convenient, please replace `$request = \Drupal::request()` with `$this->requestStack->getCurrentRequest()`. Using the injected `RequestStack` service follows Drupal’s dependency injection practices and improves maintainability

hi @vincent-gao 
yup agree, dependency injection would have been the right approach. The problem here is, prlp module does not have any service defined for the PrlpController and that is why I have to instantiate it here. I could have added a service in the tfa custom module referring to the PrlpController class, but tfa really shouldn't have a dependency on that module, so avoiding adding unnecessary dependency.

I have added the other two changes you have requested. Please have a look now. thanks :)

**Comment by MdNadimHossain** [comment]
> Why does this still have circleCI builds in it? Should the project be disconnected from Circle? Also, I don't think the failing GHA items ever worked?

yes and needs to be cleaned up and also from the submodules. I will do it in a separate PR.

File: modules/tide_tfa/src/Controller/TideTfaUserController.php [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+/**
+ * Custom controller to override the TfaUserControllerBase.
+ */
+class TideTfaUserController extends TfaUserControllerBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function doResetPassLogin($uid, $timestamp, $hash, $request = NULL) {
```
[the remark this answers is not in the document]
Comment (MdNadimHossain) [reply modules/tide_tfa/src/Controller/TideTfaUserController.php:37]: @sharmasahil I agree the logic is complex but as this is overriding the parent function, I left all the complex handling as it is from the parent and just modifying the part that is needed.

File: modules/tide_tfa/src/Controller/TideTfaUserController.php [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+      $request = \Drupal::request();
+    }
+
+    // Check if the PRLP module is enabled.
+    if (!\Drupal::moduleHandler()->moduleExists('prlp')) {
+      // If PRLP is not enabled, call the parent method.
+      return parent::doResetPassLogin($uid, $timestamp, $hash, $request);
+    }
+
+    // Create an instance of PrlpController.
```
[the remark this answers is not in the document]
Comment (MdNadimHossain) [reply modules/tide_tfa/src/Controller/TideTfaUserController.php:49]: @sharmasahil Yup same as Vincent's comment, I agree but the problem here is, prlp module does not have any service defined for the PrlpController and that is why I have to instantiate it here. I could have added a service in the tfa custom module referring to the PrlpController class, but tfa really shouldn't have a dependency on that module, so avoiding adding unnecessary dependency.

File: modules/tide_tfa/src/TideTfaOperation.php [hunk 3]
```diff
@@ -56,9 +56,13 @@ public static function setupTfaSettings() {
     $roles = Role::loadMultiple();
     // Initialize the $tfa_required_roles array.
     $tfa_required_roles = [];
+    // Define the roles to exclude in a variable.
+    $excluded_roles = ['authenticated', 'previewer', 'secure_file_user'];
+
     // Iterate through the roles and map the role IDs.
     foreach ($roles as $role) {
```
[the remark this answers is not in the document]
Comment (MdNadimHossain) [reply modules/tide_tfa/src/TideTfaOperation.php]: Done, thanks :)

File: modules/tide_tfa/src/Controller/TideTfaUserController.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (sharmasahil) [reply modules/tide_tfa/src/Controller/TideTfaUserController.php:37]: Thanks @MdNadimHossain

**Review by vincent-gao — APPROVED** [review]
lint! @MdNadimHossain 😃

**Review by anthony-malkoun — APPROVED** [review]
This looks fine to me. Thanks for cleaning up the other stuff.

**Review by sharmasahil — APPROVED** [review]
Review changes approved.

**Pull request merged**